### ISSUE TITLE
Fastor way for a safe reciprocal

### ISSUE THREAD

**Issue by ninnghazad** [romeric/Fastor#62]
What would be the Fastor way of doing something like this?
```
                auto mask{cell_mass == 0.0_r};
		cell_mass = 1.0_r / cell_mass;
		cell_mass(mask) = 0.0_r;
```

**Comment by romeric** [comment]
What is `cell_mass`? A scalar or a tensor?

**Comment by romeric** [comment]
Right okay my bad it must be a vector (1D tensor). I haven't implemented filtering or masking yet, but it is pretty high on my priority and should be available very soon as I am working on it right now.

The rest of your code should work fine though. Does this help?


~~~c++
Tensor<double,3> cell_mass = {3,4,5};
Tensor<bool,3> mask = cell_mass == 0.0;
cell_mass = 1.0 / cell_mass;
for (int i=0; i<3; ++i) if (mask(i)) cell_mass(i) = 0.0;
~~~

**Comment by romeric** [comment]
Ideally this should be almost a one liner. That is compute reciprocal aside from the elements that are zero

~~~c++
Tensor<bool,3> mask = cell_mass != 0.0; // notice !
cell_mass(mask) = 1. 0 / cell_mass(mask); 
~~~

**Comment by ninnghazad** [comment]
`cell_mass` currently is a `Tensor<float,16,16>`. At the time i thought it wouldn't matter, so i didn't mention it.

Right now i am doing it like this:
```
                Tensor<bool,16,16> mask{cell_mass == 0.0_r};
		cell_mass = 1.0_r / cell_mass;
		for(int32_t iy{0};iy < 16;++iy) {
			for(int32_t ix{0};ix < 16;++ix) {
				if(mask(iy,ix)) cell_mass(iy,ix) = 0.0_r;
			}
		}
```

But i was thinking there might be a way to use the mask to directly index the Tensor as not have to write that loop, like in your last example. 

Or maybe even some fused operation which would do away with the need for the extra mask Tensor by doing the masking/zeroing at a lower level. Was thinking of masked divide like _mm_mask_div_ps or something like that.

I also considered something like `cell_mass *= mask;`, but i would have to somehow convert my bool Tensor to a float Tensor first, and i would still have to keep a full mask.

**Comment by romeric** [comment]
To answer your second question regarding fusing first. Fusing is very much possible. In fact once implemented

~~~c++
A(mask) = B
~~~

will always fuse. The problematic part (and you have already thought of this with the conversion from float to bool) is generating `Tensor<bool,16,16>` in your example from the underlying SIMD comparison. I am sure you know that SIMD masks work differently and each SIMD vector needs its own mask. So in your example, if you want a vectorised code then instead of

~~~c++
Tensor<bool,16,16> mask{cell_mass == 0.0_r};
~~~

we should create 

~~~c++
Tensor<float,16,16> mask{cell_mass == 0.0_r};
~~~

because the comparison under the hood is this way (assuming AVX512)

~~~
__m512 mask = _mm512_cmp_pd(cell_mass.data(),_mm512_set1_pd(0.f), _CMP_EQ_OQ);
~~~

now `__m512` needs to be converted to `__m512i` and stored as an array of bools anyway. However this is quite costly. SIMD type conversions impact runtime. This is the reason, most libraries that do explicit SIMD don't bother vectorising masked operations (Eigen for example). 

In your example what you have is actually the most efficient way as the compiler emits an excellent vectorised code.

**Comment by ninnghazad** [comment]
Thank you for looking into this - and the explanation!

**Comment by romeric** [comment]
Masked/filtered views are now available (66dfc03) if you download the latest copy of Fastor. So now you can do things like

~~~c++
Tensor<double,2,2> a(2);
Tensor<bool,2,2> mask = {{false,true},{true,false}};

a(mask) = 2;
a(mask) = a(mask) + 2;
a(mask) /= <expression...>
.
.
. 
~~~

Note that the size and dimensions of mask should equal that of the tensor. This all fuses to simple mask operations and does not create any temporaries. I am closing this issue now. Feel free to open a new one, if you find an issue/bug.
